### PR TITLE
Now the user chooses the interface to send LLDP #31 packets. 

### PR DESCRIPTION
**Pull request by cmagnobarbosa**
This feature allows user reduce the CPU load on the switch.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -54,48 +55,56 @@ def execute(self):
...
-                    msg += '  -- LLDP PacketOut --\n'
-                    msg += '  Ethernet: eth_type (%s) | src (%s) | dst (%s)\n'
-                    msg += '    LLDP: Switch (%s) | port (%s)'
-
-                    log.debug(msg, switch.connection.address, switch.dpid,
-                              switch.interfaces, ethernet.ether_type,
-                              ethernet.source, ethernet.destination,
-                              switch.dpid, interface.port_number)
+                # Interface marked to receive lldp package
+                if interface.receive_lldp:
```
Comment (hdiogenes) [inline main.py]:
```
if not interface.receive_lldp:
    continue
```

File: main.py [hunk 2]
```diff
@@ -54,48 +55,56 @@ def execute(self):
...
+                    ethernet.destination = constants.LLDP_MULTICAST_MAC
+                    ethernet.data = lldp.pack()
+                    # self.vlan_id == None will result in a pack with no VLAN.
+                    ethernet.vlans.append(VLAN(vid=self.vlan_id))
+
+                    packet_out = self._build_lldp_packet_out(
+                                        of_version,
+                                        interface.port_number, ethernet.pack())
+
+                    if packet_out is not None:
```
Comment (hdiogenes) [inline main.py]:
```
if packet_out is None:
    continue
```

File: main.py [hunk 3]
```diff
@@ -283,3 +292,71 @@ def _unpack_non_empty(desired_class, data):
...
+            interfaces = interfaces + interface_list
+        return interfaces
+
+    @staticmethod
+    def as_dict(interfaces):
+        """Return a dict of interfaces."""
+        interfaces_dic = {}
+        for interface in interfaces:
+            interfaces_dic[interface.id] = interface
+        return interfaces_dic
```
Comment (hdiogenes) [inline main.py]: Replace with a dict comprehension.

File: main.py [hunk 4]
```diff
@@ -283,3 +292,71 @@ def _unpack_non_empty(desired_class, data):
...
+
+    def _get_lldp_interfaces(self):
+        """Get interfaces enabled to receive the LLDP package."""
+        enabled_receive_lldp = []
+        interfaces = self._get_interfaces()
+        for interface in interfaces:
+            if interface.receive_lldp:
+                enabled_receive_lldp.append(interface.id)
+        return enabled_receive_lldp
```
Comment (hdiogenes) [inline main.py]: Maybe replace with a list comprehension?

**Review by hdiogenes — CHANGES_REQUESTED** [review]

File: main.py (hunk 1 above)
Comment (cmagnobarbosa) [reply main.py]: Fixed.

File: main.py (hunk 4 above)
Comment (cmagnobarbosa) [reply main.py]: The code has been updated. List comprehension has been added.

File: main.py (hunk 3 above)
Comment (hdiogenes) [reply main.py]: The point of using a dict comprehension is that it can be just one line.

File: main.py [hunk 5]
```diff
@@ -283,3 +294,127 @@ def _unpack_non_empty(desired_class, data):
...
+        interfaces_dic = {}
+        interfaces_dic = {interface.id: interface for interface in interfaces}
+        return interfaces_dic
+
+    def _get_lldp_interfaces(self):
+        """Get interfaces enabled to receive the LLDP package."""
+        enabled_lldp = []
+        interfaces = self._get_interfaces()
+        enabled_lldp = [inter.id for inter in interfaces if inter.lldp]
+        return enabled_lldp
```
Comment (hdiogenes) [inline main.py]: List comprehension can/should be just one line too.

File: main.py [hunk 6]
```diff
@@ -283,3 +294,127 @@ def _unpack_non_empty(desired_class, data):
...
         return obj
+
+    @staticmethod
+    def _get_data(req):
+        """Get request data."""
+        data = req.get_json()
+        interface_list = []
+        if data:
+            interface_list = data.get('interfaces')
+        return interface_list
```
Comment (hdiogenes) [inline main.py]: `return data.get('interfaces', [])`

File: main.py [hunk 7]
```diff
@@ -283,3 +294,127 @@ def _unpack_non_empty(desired_class, data):
...
+    @rest('v1/interfaces', methods=['GET'])
+    def get_lldp_interfaces(self):
+        """Return of the interfaces available to receive the LLDP package."""
+        result = {"interfaces": self._get_lldp_interfaces()}
+        return jsonify(result), 200
+
+    @rest('v1/interfaces/disable', methods=['POST'])
+    def disable_lldp(self):
+        """Disables an interface to receive an LLDP package."""
+        interfaces_id = self._get_data(request)
```
Comment (hdiogenes) [inline main.py]: s/interfaces_id/interface_ids/

File: main.py [hunk 8]
```diff
@@ -283,3 +294,127 @@ def _unpack_non_empty(desired_class, data):
...
+        if not error_list:
+            return jsonify("All interface passed has been disabled."), 200
+
+        # Return a list of interfaces has not disabled.
+        # Other interfaces passed has enabled.
+        msg_partial_enabled = ""
+        msg_partial_enabled += "Some interfaces have not been disabled. "
+        msg_partial_enabled += "Interface list not disabled:"
+        return jsonify({msg_partial_enabled:
+                        error_list}), 400
```
Comment (hdiogenes) [inline main.py:346]: Remove lines 362, 363 and 365:

```
        # Return a list of interfaces that couldn't be disabled.
        msg_partial_enabled = "Some interfaces have not been disabled: "
```

It would also be nice to add *why* those interfaces could not be disabled.

File: main.py [hunk 9]
```diff
@@ -283,3 +294,127 @@ def _unpack_non_empty(desired_class, data):
...
+        if not error_list:
+            return jsonify("All interface passed has been enabled."), 200
+
+        # Return a list of interfaces has not enabled.
+        # The other interfaces have been activated.
+        msg_partial_enabled = ""
+        msg_partial_enabled += "Some interfaces have not been enabled. "
+        msg_partial_enabled += "Interface list not enabled:"
+        return jsonify({msg_partial_enabled:
+                        error_list}), 400
```
Comment (hdiogenes) [inline main.py:371]:
```
        # Return a list of interfaces that could not be enabled.
        # The other interfaces were enabled.
        msg_partial_enabled = "Some interfaces have not been enabled: "
```

It would also be nice to add why those interfaces could not be enabled.

File: main.py [hunk 10]
```diff
@@ -283,3 +294,127 @@ def _unpack_non_empty(desired_class, data):
...
+                interface.lldp = True
+            else:
+                # List of interfaces that were not activated.
+                error_list.append(interface_id)
+        if not error_list:
+            return jsonify("All interface passed has been enabled."), 200
+
+        # Return a list of interfaces has not enabled.
+        # The other interfaces have been activated.
+        msg_partial_enabled = ""
```
Comment (hdiogenes) [inline main.py]: Is there a need to "init" the variable like this?

File: main.py [hunk 11]
```diff
@@ -283,3 +294,127 @@ def _unpack_non_empty(desired_class, data):
...
+        result = {"interfaces": self._get_lldp_interfaces()}
+        return jsonify(result), 200
+
+    @rest('v1/interfaces/disable', methods=['POST'])
+    def disable_lldp(self):
+        """Disables an interface to receive an LLDP package."""
+        interfaces_id = self._get_data(request)
+        error_list = []
+        if interfaces_id is None:
+            return jsonify("Bad format."), 401
```
Comment (hdiogenes) [inline main.py]: Instead of checking one by one, filter out any null values.

`interface_ids = filter(None, interface_ids)`

File: main.py [hunk 12]
```diff
@@ -283,3 +294,127 @@ def _unpack_non_empty(desired_class, data):
...
+        interfaces = self._get_interfaces_dict(interfaces)
+        for interface_id in interfaces_id:
+            if not interface_id:
+                continue
+            if interfaces.get(interface_id):
+                interface = interfaces[interface_id]
+                interface.lldp = False
+            else:
+                # List of interfaces that were not activated.
+                error_list.append(interface_id)
```
Comment (hdiogenes) [inline main.py]: Move this comment to when the variable is defined:
`# List of interfaces that were not activated.`

```
        interfaces = self._get_interfaces_dict(interfaces)
        for id_ in interface_ids:
            interface = interfaces.get(id_)
            if interface:
                interface.lldp = False
            else:
                error_list.append(id_)
```

File: main.py [hunk 13]
```diff
@@ -283,3 +294,127 @@ def _unpack_non_empty(desired_class, data):
...
+            if not interface_id:
+                continue
+            if interfaces.get(interface_id):
+                interface = interfaces[interface_id]
+                interface.lldp = True
+            else:
+                # List of interfaces that were not activated.
+                error_list.append(interface_id)
+        if not error_list:
+            return jsonify("All interface passed has been enabled."), 200
```
Comment (hdiogenes) [inline main.py]: For `enable_lldp`, make the same refactoring as done in `disable_lldp`.

**Review by hdiogenes — CHANGES_REQUESTED** [review]

File: main.py [hunk 14]
```diff
@@ -68,34 +72,41 @@ def execute(self):
                 ethernet.source = interface.address
                 ethernet.destination = constants.LLDP_MULTICAST_MAC
                 ethernet.data = lldp.pack()
-                # self.vlan_id == None will result in a packet with no VLAN.
+                # self.vlan_id == None will result in a pack with no VLAN.
```
Comment (hdiogenes) [inline main.py]: revert this change

File: main.py [hunk 15]
```diff
@@ -68,34 +72,41 @@ def execute(self):
...
-                              switch.interfaces, ethernet.ether_type,
-                              ethernet.source, ethernet.destination,
-                              switch.dpid, interface.port_number)
+                event_out = KytosEvent(
+                    name='kytos/of_lldp.messages.out.ofpt_packet_out',
+                    content={
+                            'destination': switch.connection,
+                            'message': packet_out})
+                self.controller.buffers.msg_out.put(event_out)
+                # print(f"Sending package to interface {interface.id}")
```
Comment (hdiogenes) [inline main.py]: Remove `print` comment

File: main.py [hunk 16]
```diff
@@ -54,7 +55,10 @@ def execute(self):
 
             interfaces = list(switch.interfaces.values())
             for interface in interfaces:
-
+                # Interface marked to receive lldp package
+                if not interface.lldp:
+                    # print(f"Disabled LLDP package {interface.id}")
```
Comment (hdiogenes) [inline main.py]: remove print comment

s/package/packet/

Comment (cmagnobarbosa) [reply main.py]: All requested changes have been implemented.

File: main.py [hunk 17]
```diff
@@ -283,3 +292,102 @@ def _unpack_non_empty(desired_class, data):
...
+        """Return a dict of interfaces."""
+        return {inter.id: inter for inter in interfaces}
+
+    def _get_lldp_interfaces(self):
+        """Get interfaces enabled to receive the LLDP packet."""
+        return [inter.id for inter in self._get_interfaces() if inter.lldp]
+
+    @rest('v1/interfaces', methods=['GET'])
+    def get_lldp_interfaces(self):
+        """Return of the interfaces available to receive the LLDP packet."""
```
Comment (hdiogenes) [inline main.py]: Return all the interfaces that have LLDP traffic enabled.

File: main.py [hunk 18]
```diff
@@ -283,3 +292,102 @@ def _unpack_non_empty(desired_class, data):
...
+        return jsonify({"interfaces": self._get_lldp_interfaces()}), 200
+
+    @rest('v1/interfaces/disable', methods=['POST'])
+    def disable_lldp(self):
+        """Disables an interface to receive an LLDP packet."""
```
Comment (hdiogenes) [inline main.py]: s/an LLDP packet/LLDP packets/

**Review by hdiogenes — CHANGES_REQUESTED** [review]

File: main.py [hunk 19]
```diff
@@ -283,3 +292,102 @@ def _unpack_non_empty(desired_class, data):
...
+            interfaces += list(switch.interfaces.values())
+        return interfaces
+
+    @staticmethod
+    def _get_interfaces_dict(interfaces):
+        """Return a dict of interfaces."""
+        return {inter.id: inter for inter in interfaces}
+
+    def _get_lldp_interfaces(self):
+        """Get interfaces enabled to receive the LLDP packet."""
```
Comment (hdiogenes) [inline main.py]: s/the LLDP packet/LLDP packets/

File: main.py [hunk 20]
```diff
@@ -283,3 +292,102 @@ def _unpack_non_empty(desired_class, data):
...
+        """Get all interfaces."""
+        interfaces = []
+        for switch in list(self.controller.switches.values()):
+            interfaces += list(switch.interfaces.values())
+        return interfaces
+
+    @staticmethod
+    def _get_interfaces_dict(interfaces):
+        """Return a dict of interfaces."""
+        return {inter.id: inter for inter in interfaces}
```
Comment (hdiogenes) [inline main.py:312]: Much better! :)

File: main.py [hunk 21]
```diff
@@ -283,3 +292,102 @@ def _unpack_non_empty(desired_class, data):
...
+        interfaces = self._get_interfaces_dict(interfaces)
+        for id_ in interface_ids:
+            interface = interfaces.get(id_)
+            if interface:
+                interface.lldp = False
+            else:
+                error_list.append(id_)
+        if not error_list:
+            return jsonify(
+                "Every valid interface passed has been disabled."), 200
```
Comment (hdiogenes) [inline main.py]: All the requested interfaces have been disabled.

**Review by ajoaoff — APPROVED** [review]
